go-col-name using palette

midichef opened the pull request:
This PR adds a command `go-col-name`, which lets the user search for a column by its name. It uses the great new input palette so it gives better visual feedback than `go-col-regex`. To make it easy to test, for now it's bound to `gc`.

One problem with it is the matching characters are only shown when they are all lowercase.
To see this problem, `vd ~vdsrc/sample_data/sample.tsv`
then `gc` and see the matching characters are shown when you type `nits` but not when you type `Units`.

As an experiment, I also made a version that shows the contents of the column next to its name, for the cursor's row.
To get that working, I had to change `fuzzymatch()`.  Right now `fuzzymatch()` searches all keys in `haystack`. I changed it to ignore any keys that start with `'_'`. The function description mentions "each non-_key", so maybe skipping such keys is how it's meant to work, and just hasn't been implemented yet?
https://github.com/saulpw/visidata/blob/c209e6ec105d72c030a768ebe41e0d425038af33/visidata/fuzzymatch.py#L370

midichef committed:
[movement] add go-col-name

midichef committed:
[cmdpalette-] do not fuzzymatch keys that start with _

saulpw commented:
@zormit Is this how this was supposed to work anyway?

> The function description mentions "each non-_key", so maybe skipping such keys is how it's meant to work, and just hasn't been implemented yet?

zormit commented:
> @zormit Is this how this was supposed to work anyway?
> 
> > The function description mentions "each non-_key", so maybe skipping such keys is how it's meant to work, and just hasn't been implemented yet?

My honest answer after looking into this just a little bit: I'm not sure, but it sounds plausible? :D I think I haven't looked into this for too long and also this whole wrapper function including its comment seems to be from your feather 😅

saulpw commented:
Okay, let's take the `_` fix at least, I think that makes sense.

It seems like this command is useful but could use some more work; let's move it to `visidata/features/go_col.py` (perhaps along with `go-col-regex`), and then we can be more lenient with the level of polish.  Also let's structure it like `go-col-regex`, such that it assigns to `cursorVisibleColIndex` in the execstr.

saulpw commented on visidata/movement.py:
```diff
@@ -102,6 +102,52 @@ def visibleWidth(self):
+            #todo:  does not show disp_note_none for None
+            r += row._cursor_cell
+        else:
+            r += f'{name}'
```
This can just be `r += name`.  Same for another one above.

saulpw reviewed (changes requested):
With a few changes we can take this.

midichef commented:
Okay I've made the requested changes. This PR is ready to go.

A couple more changes since the last review: I make cmdpalette catch the case of 0 items to choose from. And I took away the `gc` keystroke binding for `go-col-name`. I don't think the `g` prefix makes sense for it. It doesn't fit into the Visidata philosophy for keystrokes of `_` vs`z_` vs `g_`.

anjakefala commented on visidata/movement.py:
```diff
@@ -114,7 +101,6 @@ def visibleWidth(self):
 Sheet.addCommand(None, 'go-rightmost', 'sheet.leftVisibleColIndex = len(visibleCols)-1; pageLeft(); sheet.cursorVisibleColIndex = len(visibleCols)-1', 'go all the way to the right of sheet')
 
-Sheet.addCommand('c', 'go-col-regex', 'sheet.cursorVisibleColIndex=nextColRegex(inputRegex("column name regex: ", type="regex-col", defaultLast=True))', 'go to next column with name matching regex')
 Sheet.addCommand('zc', 'go-col-number', 'sheet.cursorVisibleColIndex = int(input("move to column number: "))', 'go to given column number (0-based)')
```
I noticed in your commit message that you mention this one was being moved, too. Did you intend to move it?

midichef replied:
At first I meant to move it. Later I decided against it and forgot to change the message. I figured it belonged next to `go-row-number`, in `movement.py`. Now that I see your menu changes, the move seems better after all. Putting it in `go_col.py` keeps it with the group of commands for `Menu> Column > Goto`.

Could you squash these last few commits for the move? Thank you.

anjakefala commented:
@saulpw Ready for review
